### PR TITLE
Internal API for retrying HTTP requests

### PR DESCRIPTION
**Pull request by hiranya911**
The existing `HttpClient` retries all HTTP requests once that are failing due to connection timeout and reset errors. We would like to further extend this retries support to meet the following new requirements:

1. Ability to retry requests on specific HTTP error codes (e.g. 503)
2. Ability to delay subsequent retries according to the `Retry-After` header sent by the server
3. Ability to use exponential backoff
4. Ability to configure retries for individual services in the SDK

In order to meet all the above requirements, this PR introduces a new internal interface called `RetryConfig`. The `HttpClient` optionally takes a `RetryConfig` as a constructor parameter, and retries failing requests based on that. 

For the moment, we use the following as the default `RetryConfig` in order to retain the existing behavior:

```
const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxRetries: 1,
  ioErrorCodes: ['ECONNRESET', 'ETIMEDOUT'],
  maxDelayInMillis: 60 * 1000,
};
```

We can change this to something more sophisticated in the future (in other Admin SDKs we retry up to 4 times, and also retry on 500 and 503 HTTP errors).

### CODE REVIEW

File: src/utils/api-request.ts [hunk 1]
```diff
@@ -218,6 +286,75 @@ export class HttpClient {
...
+    const date = new Date(retryAfter);
+    if (!isNaN(date.getTime())) {
+      return date.getTime() - Date.now();
+    }
+    return -1;
+  }
+
+  private backOffDelayMillis(retryAttempts: number): number {
+    if (retryAttempts === 0) {
+      return 0;
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: This is to retain the existing behavior, and also to align with how retries are implemented in other languages/libraries. Basically we only delay retries in the event of consecutive errors. Here's a similar implementation from Python's urllib3 package: https://github.com/urllib3/urllib3/blob/master/src/urllib3/util/retry.py#L222

In general, the existing strategy of immediately retrying on the first error has worked well for us so far. This is particularly useful in environments like GCF where low-level transient errors seem to be common.

File: src/utils/api-request.ts [hunk 2]
```diff
@@ -166,8 +166,66 @@ export class HttpError extends Error {
   }
 }
 
+/**
+ * Specifies how failing HTTP requests should be retried.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: Done

File: src/utils/api-request.ts [hunk 3]
```diff
@@ -179,28 +237,38 @@ export class HttpClient {
...
     return AsyncHttpCall.invoke(config)
       .then((resp) => {
         return this.createHttpResponse(resp);
-      }).catch((err: LowLevelError) => {
-        const retryCodes = ['ECONNRESET', 'ETIMEDOUT'];
-        if (retryCodes.indexOf(err.code) !== -1 && attempts === 0) {
-          return this.sendWithRetry(config, attempts + 1);
+      })
+      .catch((err: LowLevelError) => {
+        const [delayMillis, canRetry] = this.getRetryDelayMillis(retryAttempts, err);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: I think `getRetryDelayMillis()` is the confusing one, although I think the variable names explain what's going on. In any case, I've added a jsdoc comment to the method to further explain it.

File: test/unit/utils/api-request.spec.ts [hunk 4]
```diff
@@ -636,6 +687,393 @@ describe('HttpClient', () => {
...
+      maxDelayInMillis: 10000,
+    });
+    const err = 'Error while making request: connection reset 1';
+    return client.send({
+      method: 'GET',
+      url: mockUrl,
+    }).should.eventually.be.rejectedWith(err).and.have.property('code', 'app/network-error');
+  });
+
+  it('should not retry when for error codes that are not configured', () => {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply test/unit/utils/api-request.spec.ts]: Done

File: test/unit/utils/api-request.spec.ts [hunk 5]
```diff
@@ -636,6 +687,393 @@ describe('HttpClient', () => {
...
+    return client.send({
+      method: 'GET',
+      url: mockUrl,
+    }).then((resp: HttpResponse) => {
+      expect(resp.status).to.equal(200);
+      expect(resp.headers['content-type']).to.equal('application/json');
+      expect(resp.data).to.deep.equal(respData);
+      expect(resp.isJson()).to.be.true;
+      expect(delayStub.callCount).to.equal(1);
+      expect(delayStub.args[0][0]).to.be.gt(27 * 1000).and.to.be.lte(30 * 1000);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply test/unit/utils/api-request.spec.ts]: Managed to make the assertion exact by using sinon fake timers.

**Comment by hiranya911** [comment]
Thanks @bojeil-google. Made most of the suggested changes. Ready for another look.

File: src/utils/api-request.ts [hunk 6]
```diff
@@ -166,8 +166,79 @@ export class HttpError extends Error {
...
+  maxRetries: number;
+
+  /** HTTP status codes taht should be retried. */
+  statusCodes?: number[];
+
+  /** Low-level I/O error codes that should be retried. */
+  ioErrorCodes?: string[];
+
+  /**
+   * The multiplier for exponential back off. The retry delay is calculated using the formula `(2^n) * backOffFactor`,
```
Comment (bojeil-google) [inline src/utils/api-request.ts]: Clarify that it's in seconds
The retry delay is calculated using the formula `(2^n) * backOffFactor` seconds,

File: src/utils/api-request.ts [hunk 7]
```diff
@@ -166,8 +166,79 @@ export class HttpError extends Error {
...
 }
 
+/**
+ * Specifies how failing HTTP requests should be retried.
+ */
+export interface RetryConfig {
+  /** Maximum number of times to retry a given request. */
+  maxRetries: number;
+
+  /** HTTP status codes taht should be retried. */
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: Done

File: src/utils/api-request.ts (hunk 6 above)
Comment (hiranya911) [reply src/utils/api-request.ts]: Done

**Pull request merged**